openflow13/nx_action.go

```diff
@@ -1003,3 +1005,329 @@ func NewNXActionDecTTLCntIDs(controllers uint16, ids ...uint16) *NXActionDecTTLC
 	return a
 }
+
+type NXLearnSpecHeader struct {
+	src    bool
+	dst    bool
+	output bool
+	nBits  uint16
+	length uint16
+}
+
+func (h *NXLearnSpecHeader) MarshalBinary() (data []byte, err error) {
+	data = make([]byte, h.length)
+	value := h.nBits
+	if h.src {
+		value |= 1 << LEARN_SPEC_HEADER_MATCH
+	} else {
+		value &^= 1 << LEARN_SPEC_HEADER_MATCH
+	}
+	if h.dst {
+		value |= 1 << LEARN_SPEC_HEADER_LOAD
+	} else {
+		value &^= 1 << LEARN_SPEC_HEADER_LOAD
+	}
+	if h.output {
+		value &^= 1 << LEARN_SPEC_HEADER_MATCH
+		value |= 2 << LEARN_SPEC_HEADER_LOAD
+	}
+	binary.BigEndian.PutUint16(data[0:], value)
+	return
+}
+
+func (h *NXLearnSpecHeader) UnmarshalBinary(data []byte) error {
+	if len(data) < 2 {
+		return errors.New("the []byte is too short to unmarshal a full NXLearnSpecHeader message")
+	}
+	value := binary.BigEndian.Uint16(data)
+	h.length = 2
+	h.nBits = (0xffff >> 5) & value
+	h.src = ((1 << LEARN_SPEC_HEADER_MATCH) & value) != 0
+	h.dst = ((1 << LEARN_SPEC_HEADER_LOAD) & value) != 0
+	h.output = ((2 << LEARN_SPEC_HEADER_LOAD) & value) != 0
+	return nil
+}
+
+func (h *NXLearnSpecHeader) Len() (n uint16) {
+	return h.length
+}
+
+func NewLearnHeaderMatchFromValue(nBits uint16) *NXLearnSpecHeader {
+	return &NXLearnSpecHeader{src: true, dst: false, nBits: nBits, length: 2}
+}
+
+func NewLearnHeaderMatchFromField(nBits uint16) *NXLearnSpecHeader {
+	return &NXLearnSpecHeader{src: false, dst: false, nBits: nBits, length: 2}
+}
+
+func NewLearnHeaderLoadFromValue(nBits uint16) *NXLearnSpecHeader {
+	return &NXLearnSpecHeader{src: true, dst: true, nBits: nBits, length: 2}
+}
+
+func NewLearnHeaderLoadFromField(nBits uint16) *NXLearnSpecHeader {
+	return &NXLearnSpecHeader{src: false, dst: true, nBits: nBits, length: 2}
+}
+
+func NewLearnHeaderOutputFromField(nBits uint16) *NXLearnSpecHeader {
+	return &NXLearnSpecHeader{src: false, dst: false, output: true, nBits: nBits, length: 2}
+}
+
+type NXLearnSpecField struct {
+	Field *MatchField
+	Ofs   uint16
+}
+
+func (f *NXLearnSpecField) Len() uint16 {
+	return uint16(6)
+}
+
+func (f *NXLearnSpecField) MarshalBinary() (data []byte, err error) {
+	data = make([]byte, f.Len())
+	n := 0
+	fieldValue := f.Field.MarshalHeader()
+	binary.BigEndian.PutUint32(data[n:], fieldValue)
+	n += 4
+	binary.BigEndian.PutUint16(data[n:], f.Ofs)
+	return
+}
+
+func (f *NXLearnSpecField) UnmarshalBinary(data []byte) error {
+	if len(data) < int(f.Len()) {
+		return errors.New("the []byte is too short to unmarshal a full NXLearnSpecField message")
+	}
+	f.Field = new(MatchField)
+	n := 0
+	err := f.Field.UnmarshalHeader(data[n:])
+	if err != nil {
+		return err
+	}
+	n += 4
+	f.Ofs = binary.BigEndian.Uint16(data[n:])
+	return nil
+}
+
+type NXLearnSpec struct {
+	Header   *NXLearnSpecHeader
+	SrcField *NXLearnSpecField
+	DstField *NXLearnSpecField
+	SrcValue []byte
+}
+
+func (s *NXLearnSpec) Len() uint16 {
+	length := s.Header.Len()
+	if s.Header.src {
+		length += (s.Header.nBits + 15) / 16 * 2
```

Comment:
For better readability, please rewrite the expression as 2 * (s.Header.nBits + 15) / 16. Same comment for all such instances below.

Reply:
Sure, updated.